### ISSUE TITLE
Jaws does not announce token values of aria-haspopup

### ISSUE THREAD

**Issue by Wildebrew** [FreedomScientific/standards-support#33]
### Summary
The [aria-haspopup entry in the ARIA 1.1 spec](https://www.w3.org/TR/wai-aria-1.1#aria-haspopup) allows various values (not just true/false).
aria-haspopup allows values menu, tree, grid, listbox and dialog. The attribute should be used on controls that, when activated, display those widgets.
This should be announced by Jaws, but Jaws announces all values of aria-haspopup as "has menu".

in [test file](https://codepen.io/stevef/pen/QaOXBY) navigate to the different buttons and (with the tab key or "C" because Jaws does not treat buttons with aria-haspopup as buttons, see another issue filed on this repo).

### Expected result
Jaws should announce the type of widget indicated by the value or aria-haspopup (e.g. "opens a listbox" "opens a dialog".

### Actual result

jaws only announces "button menu" for all the buttons.

### Example
(see attached file)

#### JAWS version and build number
All versions of Jaws.

 
#### Operating System and version
All versions.

#### Browser and version:
Tested in IE and Firefox, all versions (except Edge).

[aria-haspopup.txt](https://github.com/FreedomScientific/VFO-standards-support/files/1583199/aria-haspopup.txt)

**Comment by stevefaulkner** [comment]
Filed internally as Bug 98603 - Jaws does not announce token values of aria-haspopup

**Comment by scottaohara** [comment]
some additional tests https://scottaohara.github.io/tests/aria-haspopup/

**Comment by JAWS-test** [comment]
Tested links and buttons with IE 11+JAWS 2019 and got the same results as in Chrome and Firefox

**Comment by kolaps33** [comment]
I retested it now with Firefox 82.0.2 (64-bit) and Chrome Version 87.0.4280.66.
JAWS 2020.2006.12

JAWS + Chrome narrates "button menu" for button with attribure "aria-haspoup", where doesn't matter what value attribute has.

JAWS + Firefox narrates "button menu Press space to activate the menu, then navigate with arrow keys" for button with attribure "aria-haspoup", where doesn't matter what value attribute has.

Looks like issue was still not fixed. 

@stevefaulkner or @JAWS-test  please do you have some estimated when bug will be fixed? :) Thank you.

**Comment by Wildebrew** [comment]
Hi
We're closing on 2 years since I first reported this bug, still
little/no progress (ok, this sounds a bit more irritable than I meant
it to be, but it's a fact).
If there are no plans to implement support for this attribute or if
there are technical hurdles that make implementing it impossible, can
you report that back to the ARIA working group?
I'd rather have ARIA remove or deprecate roles and attributes that
have insufficient implementation support, those are bad for usability
and accessibility, I think worse than not having them at all.
I think, if after 2 years there is virtually no AT support for a role
or an attribute, that it's safe to consider it deprecated, or at least
worth revisiting.
I just saw the same response for my NVDA ticket, so it seems like the
problem is not limited to Jaws.


On 11/27/20, Milan Turon <notifications@github.com> wrote:
> I retested it now with Firefox 82.0.2 (64-bit) and Chrome Version
> 87.0.4280.66.
> JAWS 2020.2006.12
>
> JAWS + Chrome narrates "button menu" for button with attribure
> "aria-haspoup", where doesn't matter what value attribute has.
>
> JAWS + Firefox narrates "button menu Press space to activate the menu, then
> navigate with arrow keys" for button with attribure "aria-haspoup", where
> doesn't matter what value attribute has.
>
> Looks like issue was still not fixed.
>
> @stevefaulkner or @JAWS-test  please do you have some estimate when bug will
> be fixed? :) Thank you.
>
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/FreedomScientific/VFO-standards-support/issues/33#issuecomment-734898233


-- 
Work hard. Have fun. Make history.

**Comment by JAWS-test** [comment]
The value of aria-haspopup is now correctly output by JAWS (for example, when used with links). It just does not work for buttons, but that is a [specification error](https://github.com/w3c/core-aam/issues/51) and not a JAWS error

**Comment by carmacleod** [comment]
Hopefully this PR will help by making the mappings match the ARIA spec: https://github.com/w3c/core-aam/pull/86

**Comment by guyhickling** [comment]
I think it is time to remove, or at least deprecate, the new values added in  ARIA 1.1 from the aria-haspopup spec. It was a good try, but very few of the screen readers have taken them on, and they show no sign of changing in the near future. Unfortunately too many developers are using it (I used to tell them to do so myself), usually causing dialogs to be announced as menus which is confusing for blind people.

The recommended solution for developers should instead be to append "(opens dialog)" as text hidden for screen readers to the labels of buttons that open a popup dialog. That will stand the test of both time and the screen reader manufacturers!

**Comment by scottaohara** [comment]
hi @guyhickling,

feel free to post your thoughts on this ARIA issue https://github.com/w3c/aria/issues/1024, as considerations for `aria-haspopup` have been scheduled for the ARIA 1.4 milestone at this time.  (that is if you like. by me referencing this issue a link back will appear in the ARIA issue)

**Comment by Wildebrew** [comment]
There are 3 primary disadvantages of using visually hidden text over
an attribute (assuming that attribute has proper user agent support):

A. Spacing and wording issues (people often forget to put whitespace
between the end of the visible text and the visually hidden text, also
forget to put punctuation to clearly separate the label from the
instruction

B. Consistency - different people will use slightly different wording
"dialog", "a dialog", "popver" "lightbox" "overlay" etc. it will
become tricker for users to laern the pattern and set the right
expectation (then again, authors can also misuse the aria-haspopup
attribute)

C. Internationalization - attributes can be translated by screen
reader, visually hidden text needs explicit translation

 That being said, an attribute with no user agent (in this case A.T.
support) support is a useless attribute. An attribute with misleading
support (e.g. announce "has menu" for a dialog) is not only useless,
it can be damaging.

I just can't comprehend why this support hasn't been added to Jaws.
This is a very straight forward change, just announce the value of
aria-haspopup, you don't need to invent new complex announcements,
keyboard shortcuts or anything, this is only an informational
attribute that can be turned into a simple standard announcement.
I think there is big value in it if used correctly, and until it is
properly supported authors can't be trained to use it correctly.


I recommended use of the new aria-haspopup token values for awhile,
but after users started getting confused I recommend to remove it.



On 7/20/21, Guy Hickling ***@***.***> wrote:
> I think it is time to remove, or at least deprecate, the new values added in
>  ARIA 1.1 from the aria-haspopup spec. It was a good try, but very few of
> the screen readers have taken them on, and they show no sign of changing in
> the near future. Unfortunately too many developers are using it (I used to
> tell them to do so myself), usually causing dialogs to be announced as menus
> which is confusing for blind people.
>
> The recommended solution for developers should instead be to append "(opens
> dialog)" as text hidden for screen readers to the labels of buttons that
> open a popup dialog. That will stand the test of both time and the screen
> reader manufacturers!
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/FreedomScientific/VFO-standards-support/issues/33#issuecomment-883670940


-- 
Work hard. Have fun. Make history.
